Article : DuckDB et les données spatiales (#1016)
geotribu/website: content/articles/2023/2023-12-19_duckdb-donnees-spatiales.md

---
title: "DuckDB et les données spatiales"
authors:
    - Florent FOUGÈRES
categories:
    - article
comments: true
date: "2023-12-19 14:20"

> description: "Présentation de DuckDB et surtout de ses fonctions spatiales. Comment les exploiter et les mettres en liens avec des données parquet."
> image:

Reviewer: Je suggère d'utiliser le logo de DuckDB comme icône de l'article (utilisée dans la navigation de droite et l'image générée pour les cartes de partage)

```suggestion
description: "Présentation de DuckDB et surtout de ses fonctions spatiales. Comment les exploiter et les mettres en liens avec des données parquet."
icon: simple/duckdb
image:
```

> description: "Présentation de DuckDB et surtout de ses fonctions spatiales. Comment les exploiter et les mettres en liens avec des données parquet."
> image: https://cdn.icon-icons.com/icons2/3914/PNG/512/duckdb_logo_icon_248992.png

Reviewer: Ré-ajout de l'icône et suppression du logo car inadapté à l'image d'en-tête

```suggestion
description: "Présentation de DuckDB et surtout de ses fonctions spatiales. Comment les exploiter et les mettres en liens avec des données parquet."
icon: simple/duckdb
image:
```

---

> title: "DuckDB et les données spatiales"

Reviewer: pas besoin de double-quoter sauf s'il y a un caractère spécial au regard du YAML

```suggestion
title: DuckDB et les données spatiales
```

authors:
    - Florent FOUGÈRES
categories:
    - article
comments: true
date: "2023-12-19 14:20"

> description: "Présentation de DuckDB et surtout de ses fonctions spatiales. Comment les exploiter et les mettres en liens avec des données parquet."

Reviewer: ```suggestion
description: Présentation de DuckDB et surtout de ses fonctions spatiales. Comment les exploiter et les mettre en lien avec des données (géo)parquet.
```

icon: simple/duckdb
image:
license: default
robots: index, follow
tags:

>     - database

Reviewer: ```suggestion
    - base de données
```

>     - duckdb

Reviewer: ```suggestion
    - DuckDB
```

    - geoparquet
    - geos
    - parquet

>     - spatial

Reviewer: ```suggestion
```

## DuckDB c’est quoi ?

[DuckDB](https://duckdb.org/) est un SGBD (système de gestion de base de données) relationnel principalement écrit en C++ et [open source](https://github.com/duckdb/duckdb) publiée sous licence MIT. Le projet est assez récent, il a débuté en 2018, mais le projet vit beaucoup et les releases sont fréquentes, notamment grâce à une communauté très active (12 800 étoiles sur Github). Actuellement le projet est en version 0.9.1.

La particularité de ce système de base de données est qu'à l’instar des bases SQLite ou GeoPackage, les bases de données sont des fichiers portables. Ce qui permet un échange simplifié. Cependant, un des inconvénients à ce jour est le non-support de la rétrocompatibilité des bases entre les versions de DuckDB. Exemple, si quelqu’un me donne une base construite avec DuckDB 0.8.0 et que je dispose des outils de la dernière version, à savoir 0.9.1, je ne pourrais pas lire et utiliser cette base de données.

## Les performances

> Au niveau des performances, DuckDB est particulièrement adapté pour traiter avec une grande efficacité les gros volumes de données. La ou DuckDB se démarque, c’est que contrairement au système de base de données connu comme PostgreSQL ou MySQL qui ont un système de traitement par ligne, DuckDB utilise un système d’architecture en colonne et donc selon [le site officiel de DuckDB](https://duckdb.org/why_duckdb.html#duckdbisfast) :

Reviewer 2: ```suggestion
Au niveau des performances, DuckDB est particulièrement adapté pour traiter avec une grande efficacité les gros volumes de données. Au niveau des performances, DuckDB est particulièrement adapté pour traiter avec une grande efficacité les gros volumes de données. DuckDB se distingue par son approche différente : contrairement à des systèmes de base de données tels que PostgreSQL ou MySQL qui adoptent un traitement par ligne, DuckDB repose sur une architecture en colonne. Selon [le site officiel de DuckDB](https://duckdb.org/why_duckdb.html#duckdbisfast) :
```

Proposition de tournure de phrase plus légère, selon moi.

Reviewer 2: Ça me manque presqu'une explication de la différence entre colonne et ligne. Ça change quoi ? Pourquoi "c'est mieux" ? Dans quel cadre ?

> “Un grand lot de valeurs est traité en une seule opération”

> Concrètement, cela permet donc de traiter uniquement les colonnes utiles à la requête et donc d’accélérer le temps de réponse.

Reviewer 2: ```suggestion
Concrètement, cela permet donc de traiter uniquement les colonnes utiles à la requête, et ainsi accélérer le temps de réponse.```

Ajout d'une virgule avant le "et" et synonyme de "donc" qui était deux fois dans la phrase.

## Comment l’utiliser

> Tout d’abord le support est mufti plateforme, ce que je vais dire est valable pour Windows, Linux et MacOS. DuckDB est utilisable de plusieurs façons et [les installations sont bien documentées](https://duckdb.org/docs/installation/). Il y a tout d’abord un CLI si vous êtes un adepte du terminal. Mais également de nombreux langages parmi lesquelles Python, C++, R, Node.js ou encore Rust. Si vous n’êtes pas très familier avec ces environnements et que vous êtes plus à l’aise avec les gestionnaires de base de données avec interface graphique, [DBeaver](https://dbeaver.io/) propose le [support](https://duckdb.org/docs/guides/sql_editors/dbeaver.html) des bases de données DuckDB.

Reviewer 2: ```suggestion
Tout d’abord, le support est multiplateforme ; ce que j'écris est valable pour Windows, Linux et MacOS. DuckDB est utilisable de plusieurs façons et [les installations sont bien documentées](https://duckdb.org/docs/installation/). Il y a tout d’abord un CLI si vous êtes un adepte du terminal. Mais également de nombreux langages, parmi lesquels Python, C++, R, Node.js ou encore Rust. Si vous n’êtes pas très familier avec ces environnements et que vous êtes plus à l’aise avec les gestionnaires de base de données avec interface graphique, [DBeaver](https://dbeaver.io/) propose le [support](https://duckdb.org/docs/guides/sql_editors/dbeaver.html) des bases de données DuckDB.
```

Bon, on n'est pas soumis à la règle des 3 marques. Tant qu'à faire, soit en ajouter un ou deux autres comme DataGrip de JetBrains, soit dire que c'est un parmi (tant) d'autres.

Author: DataGrip ajouté :)

> Bien sûr, au-delà de la manière donc vous utilisez DuckDB, le langage pour utiliser DuckDB est le SQL. Dans la suite dans cet article, lors des exemples pratiques, je détaillerais les commandes en utilisant le CLI ou bien le paquet Python.

Reviewer 2: ```suggestion
Bien sûr, au-delà de la manière dont vous utilisez DuckDB, le langage pour utiliser DuckDB est le SQL. Dans la suite de cet article, lors des exemples pratiques, je détaillerais les commandes en utilisant le CLI ou bien le paquet Python.
```

## Les fonctions spatiales

> Les fonctions spatiales de DuckDB sont une [extension](https://duckdb.org/docs/extensions/spatial.html). La plupart des fonctions spatiales sont issus de la librairie [GEOS](https://libgeos.org/). Seulement certaines sont implémentées nativement dans le cœur de DuckDB. Si vous êtes un habitué des fonctions spatiales de PostGIS vous ne serez pas dépaysé en utilisant les fonctions spatiales du canard, la syntaxe et le nom des fonctions est extrêmement proche.

Reviewer 2: ```suggestion
Les fonctions spatiales de DuckDB sont dans une [extension](https://duckdb.org/docs/extensions/spatial.html). La plupart de ces fonctions sont issues de la librairie [GEOS](https://libgeos.org/). Seulement certaines sont implémentées nativement dans le cœur de DuckDB. Si vous êtes un habitué des fonctions spatiales de PostGIS vous ne serez pas dépaysé en utilisant les fonctions spatiales du canard : la syntaxe et le nom des fonctions sont extrêmement proches.
```

> Une bonne soixantaine de fonctions spatiales sont disponibles parmi par exemple la star de la jointure spatiale `ST_Intersects(GEOMETRY, GEOMETRY)`

Reviewer 2: ```suggestion
Une bonne soixantaine de fonctions spatiales sont disponibles, parmi, par exemple, la star de la jointure spatiale `ST_Intersects(GEOMETRY, GEOMETRY)`
```

> Ces fonctions spatiales ne sont donc pas nativement présente. Il faut donc la première fois faire un `INSTALL spatial` pour installer cette extension. Pour faire une analogie assez connue du monde de la géomatique (ou des sigistes à votre guise) c’est un peu l’équivalent d’un `CREATE EXTENSION postgis`dans PostgreSQL pour obtenir l’extension spatial PostGIS.

Reviewer 2: ```suggestion
Ces fonctions spatiales ne sont donc pas nativement présentes. Par conséquent, vous devez faire un `INSTALL spatial` pour installer cette extension. Pour faire une analogie assez connue du monde de la géomatique (ou des sigistes à votre guise) c’est un peu l’équivalent d’un `CREATE EXTENSION postgis` dans PostgreSQL pour obtenir l’extension spatiale PostGIS.```

Puis lors de chaque utilisation sur une base de données, il faut charger cette extension via un `LOAD spatial`

## Les formats de données spatiales pris en charge

> Au niveau des formats, là aussi, il y a du choix. Une cinquantaine de formats sont supportés pour lire ou importer des données spatiales. On trouve par exemples les classiques Shapefile, GeoJSON, GPX ou encore KML. Mais aussi des formats de base de données spatiales comme Spatialite ou GeoPackage. Pour finir sur les formats, si vous êtes un OSM addict, le support est encore en mode expérimental, mais DuckDB est capable de lire et intégré des fichiers [PBF](https://wiki.openstreetmap.org/wiki/PBF_Format) (`.osm.pbf`) via une fonction nommée `ST_ReadOSM()`.

Reviewer 2: ```suggestion
Au niveau des formats, là aussi, il y a du choix. Une cinquantaine de formats sont supportés pour lire ou importer des données spatiales. On trouve, par exemple, les classiques Shapefile, GeoJSON, GPX ou encore KML. Mais également des formats de base de données spatiales comme Spatialite ou GeoPackage. Pour finir sur les formats, si vous êtes un OSM addict, le support est encore en mode expérimental, mais DuckDB est capable de lire et intégrer des fichiers [PBF](https://wiki.openstreetmap.org/wiki/PBF_Format) (`.osm.pbf`) via une fonction nommée `ST_ReadOSM()`.
```

C'est supporté directement dans le code de DuckDB ou via GDAL ou autres ?

Author: Je ne connais pas la réponse à cette question. Il faudrait que je cherche.

## Les colonnes de géométries

> Une des particularités de colonne de géométrie sur DuckDB spatial, est qu’une colonne n’a pas de type de géométrie défini (comme sur PostGIS par exemple). Une même colonne de géométrie peut contenir aussi bien des points, des lignes, des polygones etc..

Reviewer 2: ```suggestion
Une des particularités de la colonne de géométrie sur DuckDB spatial, est qu’elle ne possède pas de géométrie définie ; comme sur PostGIS par exemple. Une même colonne de géométrie peut contenir aussi bien des points, des lignes, des polygones, etc.
```

Le retour à la mapinfo !

C'est considéré comme du GeometryCollection ?
Tu fais comment pour l'intégrer dans QGIS ? Split entre les différents types comme avec le GPX ?

Author: Si tu parles du plugin QGIS, le multi geom dans une couche n'est actuellement pas géré.

> Il y a quand même un hic à tout ça, à l’heure où j’écris ces lignes (peut-être que cela sera prochainement implémenté), une colonne de géométrie ne porte pas de système de projection. Il n’y a donc pas de fonction de reprojection. Et lors de l’export d’une couche géographique, cette couche ne portera pas de projection. Lors d’une lecture dans un SIG, QGIS par exemple, il faudra définir manuellement la projection de votre couche.

Reviewer 2: ```suggestion
Il y a quand même un hic à tout ça, à l’heure où j’écris ces lignes (peut-être que cela sera prochainement implémenté), une colonne de géométrie ne porte pas de système de projection. Il n’y a donc pas de fonction de reprojection. Lors de l’export d’une couche géographique, cette couche ne portera pas de projection. Lors d’une lecture dans un SIG, QGIS par exemple, il faudra définir manuellement la projection de votre couche.
```

## Les fichiers parquet, c'est quoi ?

> [Parquet](https://parquet.apache.org/) est un format de fichier open source poussé par la fondation Apache créé en juillet 2013. Et qui a la particularité de pouvoir stocker avec une grande efficacité des données en utilisant une architecture en colonne (tiens tiens cette histoire de colonne, ça ne vous dit pas quelque chose ?). Il est notamment utilisé pour des données “big data” … Les “parquet file” comme ils sont souvent nommés sur la toile ne sont que des fichiers utilisés pour s’échanger des données, ce n’est pas sûr lequel on va travailler.

Reviewer 2: ```suggestion
[Parquet](https://parquet.apache.org/) est un format de fichier open source poussé par la fondation Apache, créé en juillet 2013, qui a la particularité de pouvoir stocker avec une grande efficacité des données. Il utilise… une architecture en colonne. Tiens, tiens, cette histoire de colonne, ça ne vous dit pas quelque chose ? Il est notamment utilisé pour des données « big data »… Les « parquet files », comme ils sont souvent nommés sur la toile, sont uniquement des fichiers utilisés pour s’échanger des données et non pour travailler dessus.
```

> De nombreux outils existent pour traiter et requêter ce type de fichier, parmi un est particulièrement performant, et devinez quoi, il s’agit de DuckDB (quelle surprise hein ?)

Reviewer 2: ```suggestion
De nombreux outils existent pour traiter et requêter ce type de fichier, parmi lesquels un est particulièrement performant, et devinez quoi, il s’agit de DuckDB (quelle surprise hein ?)
```

Reviewer 2: J'ai l'impression que la façon de formuler dans le texte n'est pas forcément claire.

DuckDB, c'est une base de données ou un outil pour parquet ?

## Pour aller plus loin

> Sur ce [répertoire Github](https://github.com/davidgasquez/awesome-duckdb) est maintenu une liste de projets, outils, ou ressource développée autour de DuckDB. Petit coup de ❤️ pour [Harlequin](https://harlequin.sh/), qui est un IDE pour terminal destiné à l’utilisation de DuckDB et simple d'installation.

Reviewer 2: ```suggestion
Sur ce [répertoire Github](https://github.com/davidgasquez/awesome-duckdb) est maintenu une liste de projets, outils ou ressources développées autour de DuckDB. Petit coup de ❤️ pour [Harlequin](https://harlequin.sh/), qui est un IDE pour terminal destiné à l’utilisation de DuckDB et simple d'installation.```

![screenshort harlequin](https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2023/duckdb_spatial/harlequin.png){: .img-center }

## Exemple pratique

### Pré requis

> - L’exécutable DuckDB pour utiliser le CLI

Reviewer 2: ```suggestion
- L’exécutable DuckDB pour utiliser la CLI :
```

<!-- markdownlint-disable MD040 -->
<!-- termynal -->

```sh
v0.9.2 3c695d7ba9
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D
```

<!-- markdownlint-enable MD040 -->

<!-- markdownlint-disable MD046 -->

ou

> - Un environnement Python avec le [paquet duckdb](https://pypi.org/project/duckdb/) d’installer (DuckDB utilise de nombreuses dépendances, il est donc conseiller d’utiliser un environnement virtuel pour éviter les conflits de dépendances)

Reviewer 2: ```suggestion
- Un environnement Python avec le [paquet duckdb](https://pypi.org/project/duckdb/) déjà installé. DuckDB utilise de nombreuses dépendances, il est donc conseillé d’utiliser un environnement virtuel pour éviter les conflits de dépendances.
```

<!-- markdownlint-disable MD040 -->
<!-- termynal -->

```sh
$ pip install duckdb
---> 100%
Installing collected packages: duckdb
Successfully installed duckdb-0.9.2
```

<!-- markdownlint-enable MD040 -->

<!-- markdownlint-disable MD046 -->

> ### Création d’une de données (ou l’ouvrir si elle existe déjà)

Reviewer 2: ```suggestion
### Création d’une de données, ou l’ouvrir si elle existe déjà
```

=== ":snake: Python”

```python
import duckdb

con = duckdb.connect("./ouverture_maps-transportation.db")

> ```

Author: Je ne comprends pas ce souci, quel est le problème ?

> ```

Reviewer: cf https://github.com/geotribu/website/pull/1016#issuecomment-1842877469

=== "▶️ CLI”

```bash
D .open ouverture_maps-transportation.db
```

### Installer puis charger l’extension spatiale

> === ":snake: Python”
> 
> ```python
> con.sql("INSTALL spatial; LOAD spatial ; LOAD httpfs ;")
> ```
> 
> === "▶️ CLI”
> 
> ```bash
> D INSTALL spatial ;
> D LOAD spatial ;
> D LOAD httpfs ;
> ```

Reviewer: ````suggestion
=== ":snake: Python”

    ```python
    con.sql("INSTALL spatial; LOAD spatial ; LOAD httpfs ;")
    ```

=== "▶️ CLI”

    ```sh
    D INSTALL spatial ;
    D LOAD spatial ;
    D LOAD httpfs ;
    ```
````

### Importer un CSV et créer la géométrie

> La fonction `read_csv_auto` nous permet de pouvoir importer un CSV sans avoir à créer la table au préalable, cette fonction détecte automatiquement la structure du CSV.

Reviewer 2: ```suggestion
La fonction `read_csv_auto` nous permet de pouvoir importer un CSV sans avoir à créer la table au préalable. Cette fonction détecte automatiquement la structure du CSV.
```

> === "▶️ CLI”

Reviewer: ```suggestion
=== ":material-console: CLI”
```

> === "▶️ CLI”

Reviewer: ```suggestion
=== ":material-console: CLI”
```

> === ":snake: Python”
> 
> ```python
> con.sql("CREATE TABLE airports AS FROM read_csv_auto('[https://davidmegginson.github.io/ourairports-data/airports.csv](https://davidmegginson.github.io/ourairports-data/airports.csv)', HEADER=True, DELIM=',') ;")
> con.sql("ALTER TABLE airports ADD COLUMN the_geom GEOMETRY ;")
> con.sql("UPDATE airports SET the_geom = ST_POINT(longitude_deg, latitude_deg) ;")
> ```
> 
> === "▶ CLI”
> 
> ```bash
> D CREATE TABLE airports AS FROM read_csv_auto('[https://davidmegginson.github.io/ourairports-data/airports.csv](https://davidmegginson.github.io/ourairports-data/airports.csv)', HEADER=True, DELIM=',') ;
> D ALTER TABLE airports ADD COLUMN the_geom GEOMETRY ;
> D UPDATE airports SET the_geom = ST_POINT(longitude_deg, latitude_deg) ;
> ```

Reviewer: Correction syntaxe

````suggestion
=== ":snake: Python”

    ```python
    con.sql("CREATE TABLE airports AS FROM read_csv_auto('[https://davidmegginson.github.io/ourairports-data/airports.csv](https://davidmegginson.github.io/ourairports-data/airports.csv)', HEADER=True, DELIM=',') ;")
    con.sql("ALTER TABLE airports ADD COLUMN the_geom GEOMETRY ;")
    con.sql("UPDATE airports SET the_geom = ST_POINT(longitude_deg, latitude_deg) ;")
    ```

=== "▶ CLI”

    ```sh
    D CREATE TABLE airports AS FROM read_csv_auto('[https://davidmegginson.github.io/ourairports-data/airports.csv](https://davidmegginson.github.io/ourairports-data/airports.csv)', HEADER=True, DELIM=',') ;
    D ALTER TABLE airports ADD COLUMN the_geom GEOMETRY ;
    D UPDATE airports SET the_geom = ST_POINT(longitude_deg, latitude_deg) ;
    ```
````

> === "▶ CLI”

Reviewer: ```suggestion
=== ":material-console: CLI”
```

### Traiter des données parquet d'Ouverture Maps avec DuckDB

Les données d’Ouverture Maps sont fournies sous forme de fichier parquet ([décrites ici](https://github.com/OvertureMaps/data#data-release-feedback)), nous allons donc importer ces données dans une base pour les consulter.

#### Importer les données dans la base

> Dans cet exemple on récupère 100 bâtiments aléatoirement (environ 1 minute de traitement chez moi)

Reviewer 2: ```suggestion
Dans cet exemple, on récupère 100 bâtiments aléatoirement ; environ une minute de traitement chez moi.
```

#### Importer les données dans la base

Dans cet exemple, on récupère 100 bâtiments aléatoirement ; environ une minute de traitement chez moi.

=== ":snake: Python”

```python
query_buildings = ("CREATE TABLE buildings AS ( "  
"SELECT type, version, CAST(updatetime as varchar) as updateTime, "
"height, numfloors as numFloors, level, class, "
"ST_GeomFromWKB(geometry) as geometry "
"FROM read_parquet('s3://overturemaps-us-west-2/release/2023-11-14-alpha.0/theme=buildings/type=*/*', hive_partitioning=1) "
"LIMIT 1 ); ")

con.sql(query_admins)
```

> === "▶️ CLI”

Reviewer: ```suggestion
=== ":material-console: CLI”
```

```bash
D. CREATE TABLE buildings AS (  
 SELECT type, version, CAST(updatetime as varchar) as updateTime,
 height, numfloors as numFloors, level, class,
 ST_GeomFromWKB(geometry) as geometry
 FROM read_parquet('s3://overturemaps-us-west-2/release/2023-11-14-alpha.0/theme=buildings/type=*/*', hive_partitioning=1)
 LIMIT 1 );
```

Dans cet autre exemple, on récupère les bâtiments d’une partie de Manhattan en indiquant les coordonnées d’un rectangle (attention requête assez longue)

=== ":snake: Python”

```python
query_buildings = ("CREATE TABLE buildings AS ( "  
"SELECT type, version, CAST(updatetime as varchar) as updateTime, "
"height, numfloors as numFloors, level, class, "
"ST_GeomFromWKB(geometry) as geometry "
"FROM read_parquet('s3://overturemaps-us-west-2/release/2023-11-14-alpha.0/theme=buildings/type=*/*', hive_partitioning=1) "
"WHERE bbox.minx > -73.9967900 "
"AND bbox.maxx < -73.9967900 "
"AND bbox.miny > 40.7373325 "
"AND bbox.maxy < 40.7373325 ); ")

con.sql(query_admins)
```

> === "▶️ CLI”

Reviewer: ```suggestion
=== ":material-console: CLI”
```

 SELECT type, version, CAST(updatetime as varchar) as updateTime,
 height, numfloors as numFloors, level, class,
 ST_GeomFromWKB(geometry) as geometry
 FROM read_parquet('s3://overturemaps-us-west-2/release/2023-11-14-alpha.0/theme=buildings/type=*/*', hive_partitioning=1)
 WHERE bbox.minx > -73.9967900
 AND bbox.maxx < -73.9967900
 AND bbox.miny > 40.7373325
 AND bbox.maxy < 40.7373325 );
```

#### Visualiser les données dans QGIS

Pour cela, installer le plugin QGIS [QDuckDB](https://oslandia.gitlab.io/qgis/qduckdb/).

> :warning: Attention cette extension est encore expérimentales, il faut donc bien cocher ce paramètres dans le gestionnaire des extensions de QGIS.

Reviewer 2: ```suggestion
:warning: Attention, cette extension est encore expérimentale, il faut donc bien cocher ce paramètre dans le gestionnaire des extensions de QGIS.
```

![qduckdb](https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2023/duckdb_spatial/qduckdb.png)

#### Convertir les données en un GeoJSON en utilisant DuckDB

> Un des atouts de DuckDB n'est pas seulement d’intégrer des données pour les traiter dans la base. Mais cela peut servir d’outils de conversion pour des données parquets. Exemple, on me donne des données en parquet, mais je souhaite les avoirs en GeoJSON, on peut les convertir sans créer de table ni de base.

Reviewer 2: ```suggestion
Un des atouts de DuckDB n'est pas seulement d’intégrer des données pour les traiter dans la base. Mais, cela peut servir d’outils de conversion pour des données parquets. Exemple, on me donne des données en parquet. Pourtant, je souhaite les avoir en GeoJSON. Grâce à DuckDB, on peut les convertir sans créer de table ni de base.
```

=== ":snake: Python”

```python
query_export_buildings = ("COPY ( "  
"SELECT type, version, CAST(updatetime as varchar) as updateTime, "
"height, numfloors as numFloors, level, class, "
"ST_GeomFromWKB(geometry) as geometry "
"FROM read_parquet('s3://overturemaps-us-west-2/release/2023-11-14-alpha.0/theme=buildings/type=*/*', hive_partitioning=1) "
"WHERE bbox.minx > -73.9967900 "
"AND bbox.maxx < -73.9967900 "
"AND bbox.miny > 40.7373325 "
"AND bbox.maxy < 40.7373325 )  "
"TO 'new_york_buildings.geojson' "
" WITH (FORMAT GDAL, DRIVER 'GeoJSON', SRS 'EPSG:4326'); ")

con.sql(query_export_buildings)
```

=== "▶️ CLI”

```bash
D. COPY (
 SELECT type, version, CAST(updatetime as varchar) as updateTime,
 height, numfloors as numFloors, level, class,
 ST_GeomFromWKB(geometry) as geometry
 FROM read_parquet('s3://overturemaps-us-west-2/release/2023-11-14-alpha.0/theme=buildings/type=*/*', hive_partitioning=1)
 WHERE bbox.minx > -73.9967900
 AND bbox.maxx < -73.9967900
 AND bbox.miny > 40.7373325
 AND bbox.maxy < 40.7373325 )  
 TO 'new_york_buildings.geojson'
 WITH (FORMAT GDAL, DRIVER 'GeoJSON', SRS 'EPSG:4326');
```

> :bulb: Également possible d'exporter en Shapefile, pour cela il faut remplacer les deux dernières lignes par celles-ci :

Reviewer 2: ```suggestion
:bulb: Également possible d'exporter en Shapefile, pour cela, il faut remplacer les deux dernières lignes par celles-ci :
```

> === "▶️ CLI”

Reviewer: ```suggestion
=== ":material-console: CLI”
```

---
title: DuckDB et les données spatiales
authors:
    - Florent FOUGÈRES
categories:
    - article
comments: true
date: "2023-12-19 14:20"
description: Présentation de DuckDB et surtout de ses fonctions spatiales. Comment les exploiter et les mettre en lien avec des données (géo)parquet.
icon: simple/duckdb
image:
license: default
robots: index, follow
tags:
    - base de données
    - DuckDB
    - geoparquet
    - geos
    - parquet
---

# DuckDB et les données spatiales

:calendar: Date de publication initiale : 19 décembre 2023

## Introduction

> ![logo DuckDB](https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2023/duckdb_spatial/DuckDB_Logo.png){: .img-thumbnail-left }

Reviewer: Utilisation du logo carré déposé dans le bon dossier :

```suggestion
![logo DuckDB](https://cdn.geotribu.fr/img/logos-icones/logiciels_librairies/duckdb.png){: .img-thumbnail-left }
```

Si depuis quelques semaines, vous voyez passer beaucoup de choses sur des sujets comme DuckDB, les fichiers parquet ou encore les données d’Overture Maps, mais sans trop comprendre de quoi il s’agit, vous êtes au bon endroit !

[Commenter cet article :fontawesome-solid-comments:](#__comments){: .md-button }
{: align=middle }

----

## DuckDB c’est quoi ?

> [DuckDB](https://duckdb.org/) est un SGBD (système de gestion de base de données) relationnel principalement écrit en C++ et [open source](https://github.com/duckdb/duckdb) publié sous licence MIT. Le projet est assez récent, il a débuté en 2018, mais le projet vit beaucoup et les releases sont fréquentes, notamment grâce à une communauté très active (13 300 étoiles sur Github en date du 09/12/2023). Actuellement, le projet est en version 0.9.2.

Reviewer 3: Lien vers la [licence MIT](https://fr.wikipedia.org/wiki/Licence_MIT)

> [DuckDB](https://duckdb.org/) est un SGBD (système de gestion de base de données) relationnel principalement écrit en C++ et [open source](https://github.com/duckdb/duckdb) publié sous licence MIT. Le projet est assez récent, il a débuté en 2018, mais le projet vit beaucoup et les releases sont fréquentes, notamment grâce à une communauté très active (13 300 étoiles sur Github en date du 09/12/2023). Actuellement, le projet est en version 0.9.2.

Reviewer 3: ```suggestion
[DuckDB](https://duckdb.org/) est un SGBD (système de gestion de base de données) relationnel principalement écrit en C++ et [open source](https://github.com/duckdb/duckdb) publié sous[ licence MIT](https://fr.wikipedia.org/wiki/Licence_MIT). Le projet a débuté en 2018 et vit beaucoup. Il fait l'objet de releases fréquentes (13 300 étoiles sur GitHub le 09/12/23) et en est à la version 0.9.2.
```
```suggestion
[DuckDB](https://duckdb.org/) est un SGBD (système de gestion de base de données) relationnel principalement écrit en C++ et [open source](https://github.com/duckdb/duckdb) publié sous licence MIT. Le projet est assez récent, il a débuté en 2018, mais le projet vit beaucoup et les releases sont fréquentes, notamment grâce à une communauté très active (13 300 étoiles sur Github en date du 09/12/2023). Actuellement, le projet est en version 0.9.2.
```

> La particularité de ce système de base de données est qu'à l’instar des bases GeoPackage, File GeoDatabe d'ESRI ou encore Access MDB, les bases de données sont des fichiers portables. Ce qui permet un échange simplifié. Cependant, un des inconvénients à ce jour est le non-support de la rétrocompatibilité des bases entre les versions de DuckDB. Exemple, si quelqu’un me donne une base construite avec DuckDB 0.8.0 et que je dispose des outils de la dernière version, à savoir 0.9.2, je ne pourrais pas lire et utiliser cette base de données.

Reviewer 3: Proposition : Ce SGBD a la particularité d'être sous forme de **fichier portable** (comme les bases GeoPackage, File GeoDatabase d'ESRI ou encore Access MDB) ce qui simplifie les échanges. Cette portabilité présente cependant un défaut : la **non rétrocompatibilité** entre les différentes versions de DuckDB : un fichiers produit avec une version de DuckDB ne peut actuellement pas être lu avec une autre version de DuckDB.

> La particularité de ce système de base de données est qu'à l’instar des bases GeoPackage, File GeoDatabe d'ESRI ou encore Access MDB, les bases de données sont des fichiers portables. Ce qui permet un échange simplifié. Cependant, un des inconvénients à ce jour est le non-support de la rétrocompatibilité des bases entre les versions de DuckDB. Exemple, si quelqu’un me donne une base construite avec DuckDB 0.8.0 et que je dispose des outils de la dernière version, à savoir 0.9.2, je ne pourrais pas lire et utiliser cette base de données.

Reviewer 3: ```suggestion
Ce SGBD a la particularité d'être sous forme de fichier portable (comme les bases GeoPackage, File GeoDatabase d'ESRI ou encore Access MDB) ce qui simplifie les échanges. Cette portabilité présente cependant un défaut : la non rétrocompatibilité entre les différentes versions de DuckDB : un fichiers produit avec une version de DuckDB ne peut actuellement pas être lu avec une autre version de DuckDB.
```

## Les performances

> Au niveau des performances, DuckDB est particulièrement adapté pour traiter avec une grande efficacité les gros volumes de données. Au niveau des performances, DuckDB est particulièrement adapté pour traiter avec une grande efficacité les gros volumes de données. DuckDB se distingue par son approche différente : contrairement à des systèmes de base de données tels que PostgreSQL ou MySQL qui adoptent un traitement par ligne, DuckDB repose sur une architecture en colonne. Selon [le site officiel de DuckDB](https://duckdb.org/why_duckdb.html#duckdbisfast) :

Reviewer 3: Au niveau des performances, DuckDB est particulièrement adapté au traitement des gros volumes de données. Son architecture diffère d'ailleurs sensiblement de SGBDs comme PostgreSQL ou MySQL qui attaquent majoritairement les données par ligne, car DuckDB traite les données en colonnes. La finalité permise : _un grand lot de valeurs est traité en une seule opération_ rapporte le [le site officiel de DuckDB](https://duckdb.org/why_duckdb.html#duckdbisfast).

> Au niveau des performances, DuckDB est particulièrement adapté pour traiter avec une grande efficacité les gros volumes de données. Au niveau des performances, DuckDB est particulièrement adapté pour traiter avec une grande efficacité les gros volumes de données. DuckDB se distingue par son approche différente : contrairement à des systèmes de base de données tels que PostgreSQL ou MySQL qui adoptent un traitement par ligne, DuckDB repose sur une architecture en colonne. Selon [le site officiel de DuckDB](https://duckdb.org/why_duckdb.html#duckdbisfast) :

Reviewer 3: ```suggestion
Au niveau des performances, DuckDB est particulièrement adapté au traitement des gros volumes de données. Son architecture diffère d'ailleurs sensiblement de SGBDs comme PostgreSQL ou MySQL qui attaquent majoritairement les données par ligne, car DuckDB traite les données en colonnes. La finalité permise : un grand lot de valeurs est traité en une seule opération rapporte le [le site officiel de DuckDB](https://duckdb.org/why_duckdb.html#duckdbisfast) : on traite seulement les colonnes utiles à la requête et cela accélère et optimise le temps de réponse.
```

> > “Un grand lot de valeurs est traité en une seule opération"

Reviewer 3: ```suggestion
```

> Concrètement, cela permet donc de traiter uniquement les colonnes utiles à la requête, et ainsi accélérer le temps de réponse.

Reviewer 3: Concrètement, seules les colonnes utiles à la requête sont traitées et cela accélère et optimise le temps de réponse.

> Concrètement, cela permet donc de traiter uniquement les colonnes utiles à la requête, et ainsi accélérer le temps de réponse.

Reviewer 3: ```suggestion
```

## Comment l’utiliser

> Tout d’abord le support est multi-plateforme, ce que je vais dire est valable pour Windows, Linux et MacOS. DuckDB est utilisable de plusieurs façons et [les installations sont bien documentées](https://duckdb.org/docs/installation/). Il y a tout d’abord un CLI si vous êtes un adepte du terminal. Mais également de nombreux langages parmi lesquelles Python, C++, R, Node.js ou encore Rust. Si vous n’êtes pas très familier avec ces environnements et que vous êtes plus à l’aise avec les gestionnaires de base de données avec interface graphique, [DBeaver](https://dbeaver.io/) propose le [support](https://duckdb.org/docs/guides/sql_editors/dbeaver.html) des bases de données DuckDB.

Reviewer 3: Le support est multi-plateforme (Windows, Linux et MacOS) et DuckDB est utilisable de plusieurs façons via [des installations bien documentées](https://duckdb.org/docs/installation/). Il y a un CLI si vous êtes un adepte du terminal, et de nombreux langages peuvent l'utiliser (parmi lesquels Python, C++, R, Node.js ou encore Rust). Si vous êtes plus à l’aise avec les gestionnaires de base de données avec interface graphique, [DBeaver](https://dbeaver.io/) propose lui aussi le [support](https://duckdb.org/docs/guides/sql_editors/dbeaver.html) des bases de données DuckDB.

> Tout d’abord le support est multi-plateforme, ce que je vais dire est valable pour Windows, Linux et MacOS. DuckDB est utilisable de plusieurs façons et [les installations sont bien documentées](https://duckdb.org/docs/installation/). Il y a tout d’abord un CLI si vous êtes un adepte du terminal. Mais également de nombreux langages parmi lesquelles Python, C++, R, Node.js ou encore Rust. Si vous n’êtes pas très familier avec ces environnements et que vous êtes plus à l’aise avec les gestionnaires de base de données avec interface graphique, [DBeaver](https://dbeaver.io/) propose le [support](https://duckdb.org/docs/guides/sql_editors/dbeaver.html) des bases de données DuckDB.

Reviewer 3: ```suggestion
Le support est multi-plateforme (Windows, Linux et MacOS) et DuckDB est utilisable de plusieurs façons via [des installations bien documentées](https://duckdb.org/docs/installation/). Il y a un CLI si vous êtes un adepte du terminal, et de nombreux langages peuvent l'utiliser (parmi lesquels Python, C++, R, Node.js ou encore Rust). Si vous êtes plus à l’aise avec les gestionnaires de base de données avec interface graphique, [DBeaver](https://dbeaver.io/) propose lui aussi le [support](https://duckdb.org/docs/guides/sql_editors/dbeaver.html) des bases de données DuckDB.
```

> Bien sûr, au-delà de la manière dont vous utilisez DuckDB, le langage pour utiliser DuckDB est le SQL. Dans la suite de cet article, lors des exemples pratiques, je détaillerais les commandes en utilisant le CLI ou bien le paquet Python.

Reviewer 3: Enfin, indépendamment du client de votre choix, DuckDB fonctionne en SQL. Dans la suite de cet article, mes exemples reposeront sur des commandes en utilisant le CLI ou le paquet Python.

> Bien sûr, au-delà de la manière dont vous utilisez DuckDB, le langage pour utiliser DuckDB est le SQL. Dans la suite de cet article, lors des exemples pratiques, je détaillerais les commandes en utilisant le CLI ou bien le paquet Python.

Reviewer 3: ```suggestion
Enfin, indépendamment du client de votre choix, DuckDB fonctionne en SQL. Dans la suite de cet article, mes exemples reposeront sur des commandes en utilisant le CLI ou le paquet Python.
```

## Les fonctions spatiales

> Les fonctions spatiales de DuckDB sont une [extension](https://duckdb.org/docs/extensions/spatial.html). La plupart des fonctions spatiales sont issues de la librairie [GEOS](https://libgeos.org/). Néanmoins, toutes ne sont pas implémentées nativement dans le cœur de DuckDB. Si vous êtes un habitué des fonctions spatiales de PostGIS vous ne serez pas dépaysé en utilisant les fonctions spatiales du canard, la syntaxe et le nom des fonctions est extrêmement proche.

Reviewer 3: Les fonctions spatiales de DuckDB sont rassemblées dans une [extension](https://duckdb.org/docs/extensions/spatial.html), et sont pour la plupart issues de la librairie [GEOS](https://libgeos.org/). Néanmoins, toutes ne sont pas implémentées nativement dans le cœur de DuckDB. Si vous êtes un habitué des fonctions spatiales de PostGIS vous ne serez pas dépaysé en utilisant les fonctions spatiales du canard, la syntaxe et le nom des fonctions est extrêmement proche.

> Les fonctions spatiales de DuckDB sont une [extension](https://duckdb.org/docs/extensions/spatial.html). La plupart des fonctions spatiales sont issues de la librairie [GEOS](https://libgeos.org/). Néanmoins, toutes ne sont pas implémentées nativement dans le cœur de DuckDB. Si vous êtes un habitué des fonctions spatiales de PostGIS vous ne serez pas dépaysé en utilisant les fonctions spatiales du canard, la syntaxe et le nom des fonctions est extrêmement proche.

Reviewer 3: ```suggestion
Les fonctions spatiales de DuckDB sont rassemblées dans une [extension](https://duckdb.org/docs/extensions/spatial.html), et sont pour la plupart issues de la librairie [GEOS](https://libgeos.org/). Néanmoins, toutes ne sont pas implémentées nativement dans le cœur de DuckDB. Si vous êtes un habitué des fonctions spatiales de PostGIS vous ne serez pas dépaysé en utilisant les fonctions spatiales du canard, la syntaxe et le nom des fonctions est extrêmement proche.
```

> Une bonne soixantaine de fonctions spatiales sont disponibles, parmi, par exemple, la star de la jointure spatiale `ST_Intersects(GEOMETRY, GEOMETRY)`

Reviewer 3: On retrouve ainsi une bonne soixantaine de fonctions dont la **star de la jointure spatiale** `ST_Intersects(GEOMETRY, GEOMETRY)`

> Une bonne soixantaine de fonctions spatiales sont disponibles, parmi, par exemple, la star de la jointure spatiale `ST_Intersects(GEOMETRY, GEOMETRY)`

Reviewer 3: ```suggestion
On retrouve ainsi une bonne soixantaine de fonctions dont la star de la jointure spatiale ST_Intersects(GEOMETRY, GEOMETRY)
```

> Ces fonctions spatiales ne sont donc pas nativement présentes. Il faut donc la première fois faire un `INSTALL spatial` pour installer cette extension. Pour faire une analogie assez connue du monde de la géomatique (ou des sigistes à votre guise) c’est un peu l’équivalent d’un `CREATE EXTENSION postgis`dans PostgreSQL pour obtenir l’extension spatial PostGIS.

Reviewer 3: Ces fonctions ne sont donc pas natives mais s'implémentent avec un simple `INSTALL spatial`, qui équivaut (coucou les géomaticien.ne.s) à un `CREATE EXTENSION postgis` pour obtenir l’extension spatial PostGIS dans PostgreSQL.

> Ces fonctions spatiales ne sont donc pas nativement présentes. Il faut donc la première fois faire un `INSTALL spatial` pour installer cette extension. Pour faire une analogie assez connue du monde de la géomatique (ou des sigistes à votre guise) c’est un peu l’équivalent d’un `CREATE EXTENSION postgis`dans PostgreSQL pour obtenir l’extension spatial PostGIS.

Reviewer 3: ```suggestion
Ces fonctions ne sont donc pas natives mais s'implémentent avec un simple INSTALL spatial, qui équivaut (coucou les géomaticien.ne.s) à un CREATE EXTENSION postgis pour obtenir l’extension spatial PostGIS dans PostgreSQL. À chaque utilisation, cette extension s'appelle par la commande LOAD spatial.
```

> Puis lors de chaque utilisation sur une base de données, il faut charger cette extension via un `LOAD spatial`

Reviewer 3: ```suggestion
```

## Les formats de données spatiales pris en charge

> Au niveau des formats, là aussi, il y a du choix. Une cinquantaine de formats sont supportés pour lire ou importer des données spatiales. On trouve par exemple les classiques Shapefile, GeoJSON, GPX ou encore KML. Mais aussi des formats de base de données spatiales comme Spatialite ou GeoPackage. Pour finir sur les formats, si vous êtes un OSM addict, le support est encore en mode expérimental, mais DuckDB est capable de lire et intégrer des fichiers [PBF](https://wiki.openstreetmap.org/wiki/PBF_Format) (`.osm.pbf`) via une fonction nommée `ST_ReadOSM()`.

Reviewer 3: Au niveau des formats de données spatiales, beaucoup de choix également. Une cinquantaine sont supportés en lecture ou import dont les classiques Shapefile, GeoJSON, GPX ou encore KML, ainsi quedes formats de base de données spatiales comme Spatialite ou GeoPackage. Enfin, pour les OSM addicts, en mode expérimental, DuckDB est capable de lire et intégrer des fichiers [PBF](https://wiki.openstreetmap.org/wiki/PBF_Format) (`.osm.pbf`) via une fonction nommée `ST_ReadOSM()`.

> Au niveau des formats, là aussi, il y a du choix. Une cinquantaine de formats sont supportés pour lire ou importer des données spatiales. On trouve par exemple les classiques Shapefile, GeoJSON, GPX ou encore KML. Mais aussi des formats de base de données spatiales comme Spatialite ou GeoPackage. Pour finir sur les formats, si vous êtes un OSM addict, le support est encore en mode expérimental, mais DuckDB est capable de lire et intégrer des fichiers [PBF](https://wiki.openstreetmap.org/wiki/PBF_Format) (`.osm.pbf`) via une fonction nommée `ST_ReadOSM()`.

Reviewer 3: ```suggestion
Au niveau des formats de données spatiales, beaucoup de choix également. Une cinquantaine sont supportés en lecture ou import dont les classiques Shapefile, GeoJSON, GPX ou encore KML, ainsi quedes formats de base de données spatiales comme Spatialite ou GeoPackage. Enfin, pour les OSM addicts, en mode expérimental, DuckDB est capable de lire et intégrer des fichiers [PBF](https://wiki.openstreetmap.org/wiki/PBF_Format) (.osm.pbf) via une fonction nommée ST_ReadOSM().
```

> ## Les colonnes de géométries

Reviewer 3: ## Les colonnes de géométries et la non prise en charge des projections

> ## Les colonnes de géométries

Reviewer 3: ```suggestion
## Les colonnes de géométries et la non prise en charge des projections
```

> Une des particularités de colonne de géométrie sur DuckDB spatial, est qu’une colonne n’a pas de type de géométrie défini (comme sur PostGIS par exemple). Une même colonne de géométrie peut contenir aussi bien des points, des lignes, des polygones, etc..

Reviewer 3: ```suggestion
Une des particularités des colonnes de géométrie sur DuckDB spatial est l'absence de définition du type de géométrie (contrairement à PostGIS par exemple). Une même colonne de géométrie contiendra aussi bien des points, des lignes, des polygones, etc.
```

> Il y a quand même un hic à tout ça, à l’heure où j’écris ces lignes (peut-être que cela sera prochainement implémenté), il n'est pas possible de définir une projection comme contrainte pour une colonne. Il n’y a donc pas de fonction de reprojection. Et lors de l’export d’une couche géographique, cette couche ne portera pas de projection. Lors d’une lecture dans un SIG, QGIS par exemple, il faudra définir manuellement la projection de votre couche. Cela implique une certaine rigueur, car toutes vos données de la table devront donc être dans la même projection. Même si cette réflexion tombe sous le sens du point de vue "administrateur de données", les utilisateurs moins avertis peuvent avoir quelques surprises en mélangeant des projections.

Reviewer 3: ```suggestion
Il y a quand même un hic dans tout ça : la définition d'une projection comme contrainte pour une colonne n'est actuellement pas implémentée. Il n’y a pas non plus de fonction de reprojection. De même qu'une couche n'a pas de projection attribuée lors d'un export. Il vous faudra définir manuellement cette projection en dehors de DuckDB, ce qui implique une certaine rigueur : toutes vos données devront donc être dans la même projection.

Si cette réflexion tombe sous le sens du point de vue "administrateur de données", les utilisateurs moins avertis peuvent avoir quelques surprises en mélangeant des projections.
```

## Les fichiers parquet, c'est quoi ?

> 
> [Parquet](https://parquet.apache.org/) est un format de fichier open source poussé par la fondation Apache, créé en juillet 2013, qui a la particularité de pouvoir stocker avec une grande efficacité des données. Il utilise… une architecture en colonne. Tiens, tiens, cette histoire de colonne, ça ne vous dit pas quelque chose ? Il est notamment utilisé pour des données « big data »… Les « parquet files », comme ils sont souvent nommés sur la toile, sont uniquement des fichiers utilisés pour s’échanger des données et non pour travailler dessus.

Reviewer: Ajout logo du format Parquet

```suggestion

![logo Parquet](https://cdn.geotribu.fr/img/logos-icones/logiciels_librairies/parquet.png){: .img-thumbnail-left }

[Parquet](https://parquet.apache.org/) est un format de fichier open source poussé par la fondation Apache, créé en juillet 2013, qui a la particularité de pouvoir stocker avec une grande efficacité des données. Il utilise… une architecture en colonne. Tiens, tiens, cette histoire de colonne, ça ne vous dit pas quelque chose ? Il est notamment utilisé pour des données « big data »… Les « parquet files », comme ils sont souvent nommés sur la toile, sont uniquement des fichiers utilisés pour s’échanger des données et non pour travailler dessus.
```

Author: Repris dans ce [commit](1442a0950163fd446dd9902463799b63c609e2ba) pour également prendre en compte la reprise de @Gab-OPenIG

> [Parquet](https://parquet.apache.org/) est un format de fichier open source poussé par la fondation Apache, créé en juillet 2013, qui a la particularité de pouvoir stocker avec une grande efficacité des données. Il utilise… une architecture en colonne. Tiens, tiens, cette histoire de colonne, ça ne vous dit pas quelque chose ? Il est notamment utilisé pour des données « big data »… Les « parquet files », comme ils sont souvent nommés sur la toile, sont uniquement des fichiers utilisés pour s’échanger des données et non pour travailler dessus.

Reviewer 3: ```suggestion
[Parquet](https://parquet.apache.org/) est un format de fichier open source poussé par la fondation Apache depuis 2013, qui a la particularité de stocker des données avec efficacité... via une architecture en colonne 😉. Il est notamment utilisé pour le big data. À la base, ces « parquet files » étaient utilisés pour l'échange de données et non pour travailler dessus mais les schémas de compression et de codage de données qu'ils mobilisent les rendent très performants pour la gestion massive de données complexes.
```

Author: Repris dans ce [commit](1442a0950163fd446dd9902463799b63c609e2ba) pour également prendre en compte l'ajout du logo parquet proposé par @Guts

> De nombreux outils existent pour traiter et requêter ce type de fichier, parmi lesquels un est particulièrement performant, et devinez quoi, il s’agit de DuckDB (quelle surprise hein ?)

Reviewer 3: ```suggestion
```

    "WHERE bbox.minx > -73.9967900 "
    "AND bbox.maxx < -73.9967900 "
    "AND bbox.miny > 40.7373325 "
    "AND bbox.maxy < 40.7373325 ); ")

    con.sql(query_admins)
    ```

=== ":material-console: CLI"

    ```sh
    D. CREATE TABLE buildings AS (  
        SELECT type, version, CAST(updatetime as varchar) as updateTime,
            height, numfloors as numFloors, level, class,
        ST_GeomFromWKB(geometry) as geometry
        FROM read_parquet('s3://overturemaps-us-west-2/release/2023-11-14-alpha.0/theme=buildings/type=*/*', hive_partitioning=1)
        WHERE bbox.minx > -73.9967900
            AND bbox.maxx < -73.9967900
            AND bbox.miny > 40.7373325
            AND bbox.maxy < 40.7373325 );
    ```

#### Visualiser les données dans QGIS

Pour cela, installer le plugin QGIS [QDuckDB](https://oslandia.gitlab.io/qgis/qduckdb/).
:warning: Attention, cette extension est encore expérimentale, il faut donc bien cocher ce paramètre dans le gestionnaire des extensions de QGIS.

![qduckdb](https://cdn.geotribu.fr/img/articles-blog-rdp/articles/2023/duckdb_spatial/qduckdb.png){: .img-center loading=lazy }

#### Convertir les données en un GeoJSON en utilisant DuckDB

> Un des atouts de DuckDB n'est pas seulement d’intégrer des données pour les traiter dans la base. Mais, cela peut servir d’outils de conversion pour des données parquets. Exemple, on me donne des données en parquet. Pourtant, je souhaite les avoir en GeoJSON. Grâce à DuckDB, on peut les convertir sans créer de table ni de base.

Reviewer 3: ```suggestion
Un des atouts de DuckDB est qu'en plus d’intégrer des données pour les traiter en base, il peut servir d’outil de conversion pour des données parquets. Exemple : Si on me donne des données en parquet et je souhaite des GeoJSON, DuckDB peut les convertir sans créer de table ni de base !
```
